ludwig/features/audio_feature.py

```diff
@@ -0,0 +1,301 @@
+            feature_dim_symmetric = get_length_in_samp(audio_feature_dict['window_length_in_s'], sampling_rate_in_hz)
+            feature_dim = get_non_symmetric_length(feature_dim_symmetric)
+        else:
+            raise ValueError('{} is not recognized.'.format(feature_type))
+
+        return feature_dim
+
+    @staticmethod
+    def _read_audio_and_transform_to_feature(filepath, audio_feature_dict, feature_dim, max_length, padding_value, normalization_type, audio_stats):
+        """
+        :param filepath: path to the audio
+        :param audio_feature_dict: dictionary describing audio feature see default
+        :param feature_dim: dimension of each feature frame
+        :param max_length: max audio length defined by user in samples
+        """
+
+        feature_type = audio_feature_dict['type']
+        audio, sampling_rate_in_hz = soundfile.read(filepath)
+        AudioBaseFeature._update(audio_stats, audio, sampling_rate_in_hz)
+
+        if(feature_type == 'raw'):
+            audio_feature = np.expand_dims(audio, axis=-1)
+        elif(feature_type in ['stft', 'stft_phase', 'group_delay']):
+            audio_feature = np.transpose(AudioBaseFeature._get_2D_feature(audio, feature_type, audio_feature_dict, sampling_rate_in_hz))
+        else:
+            raise ValueError('{} is not recognized.'.format(feature_type))
+
+        feature_length = audio_feature.shape[0]
+        broadcast_feature_length = min(feature_length, max_length)
+        audio_feature_padded = np.full((max_length, feature_dim), padding_value, dtype=np.float32)
+        audio_feature_padded[:broadcast_feature_length, :] = audio_feature[:max_length, :]
+        return audio_feature_padded
+
+    @staticmethod
+    def _update(audio_stats, audio, sampling_rate_in_hz):
+        audio_length_in_s = audio.shape[-1] / float(sampling_rate_in_hz)
+        audio_stats['count'] += 1
+        mean = (( audio_stats['count'] - 1) * audio_stats['mean'] + audio_length_in_s ) / float(audio_stats['count'])
+        mean = calculate_incr_mean(audio_stats['count'], audio_stats['mean'], audio_length_in_s)
+        var = calculate_incr_var(audio_stats['var'], audio_stats['mean'], mean, audio_length_in_s)
+        audio_stats['mean'] = mean
+        audio_stats['var'] = var
+        audio_stats['max'] = max(audio_stats['max'], audio_length_in_s)
+        audio_stats['min'] = min(audio_stats['min'], audio_length_in_s)
+        if(audio_length_in_s > audio_stats['max_length_in_s']):
+            audio_stats['cropped'] += 1
+
+    @staticmethod
+    def _get_2D_feature(audio, feature_type, audio_feature_dict, sampling_rate_in_hz):
+        window_length_in_s = audio_feature_dict['window_length_in_s']
+        window_shift_in_s = audio_feature_dict['window_shift_in_s']
+        num_fft_points = audio_feature_dict['num_fft_points'] if 'num_fft_points' in audio_feature_dict else get_length_in_samp(window_length_in_s, sampling_rate_in_hz)
```

Comment:
Are fft points essentially frequencies, should this be num_fft_freqs or it's something else?

Reply:
changed naming to num_fft_bins (e954869). num_fft_bins describes the "resolution" of the fft transform. If num_fft_bins > window_length_in_samp the array is zero padded (see "nfft" in https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.stft.html)